pkg/controllers/node/health/controller.go

```diff
@@ -0,0 +1,125 @@
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+)
+
+// Controller for the resource
+type Controller struct {
+	clock         clock.Clock
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.CloudProvider, clock clock.Clock) *Controller {
+	return &Controller{
+		clock:         clock,
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+	}
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	return controllerruntime.NewControllerManagedBy(m).
+		Named("node.health").
+		For(&corev1.Node{}).
+		Complete(reconcile.AsReconciler(m.GetClient(), c))
+}
+
+func (c *Controller) Reconcile(ctx context.Context, node *corev1.Node) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "node.health")
+	ctx = log.IntoContext(ctx, log.FromContext(ctx).WithValues("Node", klog.KRef(node.Namespace, node.Name)))
+
+	// Validate that the node is owned by us and is not being deleted
+	nodeClaim, err := nodeutils.NodeClaimForNode(ctx, c.kubeClient, node)
+	if err != nil {
+		return reconcile.Result{}, nodeutils.IgnoreNodeClaimNotFoundError(err)
+	}
+
+	requeueTime := time.Time{}
+	foundHealthCondition := false
+	cpPolicyFound := cloudprovider.RepairPolicy{}
+	// Find a node with a condition that matches one of the unhealthy conditions defined by the cloud provider
+	// If there are multiple unhealthy status condition we will requeue based on the condition closest to its terminationDuration
+	for _, policy := range c.cloudProvider.RepairPolicies() {
+		// check the status and the type on the condition
+		nodeCondition := nodeutils.GetCondition(node, policy.ConditionType)
+		if nodeCondition.Status == policy.ConditionStatus {
+			terminationTime := nodeCondition.LastTransitionTime.Add(policy.TolerationDuration)
+			// Determine requeue time
+			if requeueTime.Before(terminationTime) {
+				requeueTime = terminationTime
+				foundHealthCondition = c.clock.Now().After(terminationTime)
+				cpPolicyFound = policy
+			}
+		}
+	}
+
+	// If the Node is unhealthy, but has not reached it's full toleration disruption
+	// requeue at the termination time of the unhealthy node
+	if !foundHealthCondition {
+		return reconcile.Result{RequeueAfter: requeueTime.Sub(c.clock.Now())}, nil
+	}
+
+	// For unhealthy past the tolerationDisruption window we can forcefully terminate the node
+	if err := c.annotateTerminationGracePeriod(ctx, nodeClaim); err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+	if err := c.kubeClient.Delete(ctx, nodeClaim); err != nil {
+		return reconcile.Result{}, err
```

Comment:
This delete could have a NotFound error -- consider adding a `client.IgnoreNotFound`